external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala

```diff
@@ -100,13 +100,14 @@ class AvroSerializer(rootCatalystType: DataType, rootAvroType: Schema, nullable:
           et, resolveNullableType(avroType.getElementType, containsNull))
         (getter, ordinal) => {
           val arrayData = getter.getArray(ordinal)
-          val result = new java.util.ArrayList[Any]
+          val len = arrayData.numElements()
+          val result = new Array[Any](len)
```

Comment:
one more improvement: if the element is primitive type, we can call `arrayData.toBoolean/Int/...Array` directly.

Reply:
I tested this out, and this doesn't help much. 

I guess the reason is the avro writer is expecting a boxed `ArrayList`, so even we call the primitive APIs, Scala will still do the auto-boxing which will not be much different than the current code.